File: school/summer2013/example_openmp/fortran/08_false_sharing_solved.f90
program false_sharing_solved
    integer(8) :: total_sum=0, local_sum(0:(TN-1)*8)
    integer i,tid, omp_get_thread_num

!$omp parallel private(tid) num_threads(TN)
    tid = omp_get_thread_num() * 8
    local_sum(tid) = 0
    !$omp do
    do i=1, N
        local_sum(tid) = local_sum(tid) + i
    end do
!$omp end parallel

    do i=0, TN-1
        total_sum = total_sum + local_sum(i*8)
    end do

    print *, 'sum=', total_sum
end

File: summer2013/openmp/c/08_false_sharing_solved.c
#include <stdio.h>
#include <omp.h>

int main()
{
	long i, sum=0, tid, local_sum[TN*8];

#pragma omp parallel private(tid) num_threads(TN)
{
	tid = omp_get_thread_num();
	local_sum[tid*8] = 0;

	#pragma omp for
	for(i=1; i<=N; i++)
		local_sum[tid*8] += i;
}
	for(i=0; i<TN; i++)
		sum += local_sum[i*8];

	printf("sum = %ld\n", sum);
}
